qiskit/primitives/statevector_sampler.py

```diff
@@ -0,0 +1,225 @@
+    qreg_indices: List[int]
+
+
+class Sampler(BaseSamplerV2):
+    """
+    Simple implementation of :class:`BaseSamplerV2` with Statevector.
+
+    :Options:
+
+        - **shots** (int) --
+          The number of shots.
+
+        - **seed** (np.random.Generator or int) --
+          Set a fixed seed or generator for the normal distribution.
+    """
+
+    _options_class = Options
+    options: Options
+
+    def __init__(self, *, options: Optional[BasePrimitiveOptions] = None):
+        """
+        Args:
+            options: Options including shots, seed.
+        """
+        if options is None:
+            self.options = Options()
+        elif isinstance(options, BasePrimitiveOptions):
+            self.options = options
+        else:
+            self.options = Options(**options)
+
+    def run(
+        self, pubs: Iterable[SamplerPubLike], shots: int | None = None
+    ) -> PrimitiveJob[PrimitiveResult[PubResult]]:
+        job: PrimitiveJob[PubResult] = PrimitiveJob(self._run, pubs, shots)
+        job._submit()
+        return job
+
+    def _run(
+        self, pubs: Iterable[SamplerPubLike], shots: int | None = None
+    ) -> PrimitiveResult[PrimitiveResult[PubResult]]:
+        if shots is None:
+            shots = self.options.shots
+        coerced_pubs = [SamplerPub.coerce(pub, shots) for pub in pubs]
+        for pub in coerced_pubs:
+            pub.validate()
+
+        results = []
+        for pub in coerced_pubs:
+            circuit, qargs, meas_info = _preprocess_circuit(pub.circuit)
+            bound_circuits = pub.parameter_values.bind_all(circuit)
+            arrays = {
+                item.creg_name: np.zeros(
+                    bound_circuits.shape + (shots, item.num_bytes), dtype=np.uint8
+                )
+                for item in meas_info
+            }
+            for index in np.ndindex(*bound_circuits.shape):
+                bound_circuit = bound_circuits[index]
+                final_state = Statevector(bound_circuit_to_instruction(bound_circuit))
+                final_state.seed(self.options.seed)
+                if qargs:
+                    samples = final_state.sample_memory(shots=shots, qargs=qargs)
+                else:
+                    samples = [""] * shots
+                samples_array = np.array(
+                    [np.fromiter(sample, dtype=np.uint8) for sample in samples]
+                )
+                for item in meas_info:
+                    ary = _samples_to_packed_array(samples_array, item.num_bits, item.qreg_indices)
+                    arrays[item.creg_name][index] = ary
+
+            data_bin_cls = make_data_bin(
+                [(item.creg_name, BitArray) for item in meas_info],
+                shape=bound_circuits.shape,
+            )
+            meas = {
+                item.creg_name: BitArray(arrays[item.creg_name], item.num_bits)
+                for item in meas_info
+            }
+            data_bin = data_bin_cls(**meas)
```

Comment:
I attempted to implement myself using just `BitArray.from_samples`, which is in-principle compatible with the output of `sample_memory`. I found it difficult to do cleanly, mainly because of the register slicing (that you are doing in `_samples_to_packed_array`). The main reason I attempted to do this was to check if our tools are easy to use, and whether or not they promote excessive copying of data. My conclusion is that we're missing at least one helper method/function along the lines of 

```python
def populate_from_samples(samples: Iterable[int | str], register_mapping: dict[str, Iterable[int]],  shape: ShapeInput) -> DataBin:
    """"Construct a new data bin from a stream of samples.
    
    Args:
        samples: An iterable of samples (like in from_samples).
        register_mapping: A mapping from register names to their positions in each bitstring.
        shape: The shape of each bit array.
        
    Returns:
        A new data bin containing one bit array for every register in the mapping.
    """
```

Since your implementation works, and we are short on time, I do not think we should prioritize refactoring in this way, but it would be nice to eventually have enough tools so that people don't have to think about uint8s in the majority (but not all!) of cases.

Reply:
Yes, the register slicing is complicated. I also think it's specific to Statevector-based Sampler because Statevector returns bitstrings aligned to qregs and we need to reorder them following measured cregs. On the other hand, backend-based Sampler may not need take care of the register slicing because backend returns bitstrings aligned to measured cregs directly. So, we might not need a public helper method/function for register slicing.